db/multi_cf_iterator.h

```diff
@@ -65,52 +69,113 @@ class MultiCfIterator : public Iterator {
       assert(b.iterator->Valid());
       int c = comparator_->Compare(a.iterator->key(), b.iterator->key());
       assert(c != 0 || a.order != b.order);
-      return c == 0 ? a.order - b.order > 0 : c > 0;
+      return c == 0 ? a.order - b.order > 0 : CompareOp()(c, 0);
     }
 
    private:
     const Comparator* comparator_;
   };
-
   const Comparator* comparator_;
   using MultiCfMinHeap =
-      BinaryHeap<MultiCfIteratorInfo, MultiCfMinHeapItemComparator>;
-  MultiCfMinHeap min_heap_;
-  // TODO: MaxHeap for Reverse Iteration
+      BinaryHeap<MultiCfIteratorInfo,
+                 MultiCfHeapItemComparator<std::greater<int>>>;
+  using MultiCfMaxHeap = BinaryHeap<MultiCfIteratorInfo,
+                                    MultiCfHeapItemComparator<std::less<int>>>;
+
+  using MultiCfIterHeap = std::variant<MultiCfMinHeap, MultiCfMaxHeap>;
+
+  MultiCfIterHeap heap_;
+
+  enum Direction : uint8_t { kForward, kReverse };
+  Direction direction_ = kForward;
+
   // TODO: Lower and Upper bounds
 
+  Iterator* current() const {
+    if (direction_ == kReverse) {
+      auto& max_heap = std::get<MultiCfMaxHeap>(heap_);
+      return max_heap.top().iterator;
+    }
+    auto& min_heap = std::get<MultiCfMinHeap>(heap_);
+    return min_heap.top().iterator;
+  }
+
   Slice key() const override {
     assert(Valid());
-    return min_heap_.top().iterator->key();
+    return current()->key();
+  }
+  Slice value() const override {
+    assert(Valid());
+    return current()->value();
+  }
+  const WideColumns& columns() const override {
+    assert(Valid());
+    return current()->columns();
+  }
```

Comment:
Not strictly related to this patch but this makes `value()` and `columns()` dependent on which CFs out of the entire set the key is present in (without indicating which CF it is), which feels counterintuitive. It might be nicer to use `value()` and `columns()` to either
a) expose the value/columns of the default column family or
b) expose the value/columns of the first column family in the list

Reply:
We know at least one internal use case that is combining column families in this way, so we expect it to be useful functionality to provide. It's only counter-intuitive if you consider Iterator::Next/value/key/etc. to be tightly coupled to the notion of compator-ordered KVs for a single CF. Looking at the definition of Iterator it seems quite abstract. For example, you can't ask it which CF the stream of KVs come from; it just provides the stream with implicit context.

> a) expose the value/columns of the default column family or

What if the default column family wasn't on the list? Is that more intuitive?

> b) expose the value/columns of the first column family in the list

In both cases, there's no consistency on whether Next() relates to single-CF value()/columns() or to multi-CF attributes.

HOWEVER, I do feel like columns() should be combining across the listed CFs, treating each column as independently shadowable across the CFs. Or at least there should be an option to get this functionality. (I don't recall where previous discussions landed on this.) For example, I might have CFs with different storage profiles, and want to change which one I'm writing a particular column to without having to change my read logic, or explicit implement the combining+shadowing myself.

Reply:
>  Looking at the definition of Iterator it seems quite abstract. For example, you can't ask it which CF the stream of KVs come from; it just provides the stream with implicit context.

Not sure I'm following this. For the regular (non-multi-CF) iterator, there is only one CF: the one passed during iteration creation. So the application has a trivial way of knowing which CF the KVs come from. Here, on the other hand, the CF information is not readily available; if key1 is present in CF2 and CF4, and key2 is present in CF1 and CF3, value() and columns() might expose data from CF2 for key1 and data from CF1 for key2, where the CFs might have completely different semantics.

> > a) expose the value/columns of the default column family or
> 
> What if the default column family wasn't on the list? Is that more intuitive?

Obviously we would document this and leave value() and columns() empty in this case. With that caveat, yes, it would be more intuitive IMHO.

> 
> > b) expose the value/columns of the first column family in the list
> 
> In both cases, there's no consistency on whether Next() relates to single-CF value()/columns() or to multi-CF attributes.

Sorry, I'm not following this consistency argument either. The way this supposed to work is that we combine data for the same key across column families, and Next takes you to the next *key*. In the above example, when the iterator is positioned on key1, attribute_groups() would return { CF2: <key1's columns from CF2>, CF4: <key1's columns for CF4>}, and  Next() would take you from key1 to key2, at which point attribute_groups() would return {CF1: <key2's columns from CF1>, CF3: <key2's columns for CF3>}. In general, when doing multi-CF iteration, I would expect applications to primarily use `attribute_groups()`, and only use `columns()` and `value()` as a shorthand for… something (see above).

> 
> HOWEVER, I do feel like columns() should be combining across the listed CFs, treating each column as independently shadowable across the CFs. Or at least there should be an option to get this functionality. (I don't recall where previous discussions landed on this.) For example, I might have CFs with different storage profiles, and want to change which one I'm writing a particular column to without having to change my read logic, or explicit implement the combining+shadowing myself.

Not in favor of this. Attribute groups are supposed to have namespace functionality - you can have e.g. a column named "timestamp" in many different attribute groups (timestamp of last login, last post, last share etc.) that have no relation to each other and are not meant to be combined.

Reply:
Related to the above, `MultiCfIterator::Next()` should advance all child iterators positioned on the current key, and similarly for `Prev()` etc.

Reply:
> Not sure I'm following this. For the regular (non-multi-CF) iterator, there is only one CF: the one passed during iteration creation.

Iterator creation is not part of the Iterator class. It's an abstract class.

> Sorry, I'm not following this consistency argument either.

CF1: key1, key3
CF2:     key2

MultiCfIterator with {CF1,CF2} is at key1, calls Next(), then key() returns what? If it's key2, what is value()? If it's key3, how do I get the attribute_groups()?

Reply:
@pdillinger in your example,  After `Next()` from key1, `key()` should return key2. `value()` should be the value of the key2 in CF2. `Next()` then will get to key3 in CF1.

`attribute_groups()` won't have much value unless keys exist in multiple CFs. In the example above, the only value it could provide is the indication of which CF the key/value came from.  
it will return the following for each key. (assuming the values are plain texts, not wide columns)

```
key_1 ==> {cf: "CF1", columns: {kDefaultColumnName: "value for key 1"}}
key_2 ==> {cf: "CF2", columns: {kDefaultColumnName: "value for key 2"}}
key_3 ==> {cf: "CF1", columns: {kDefaultColumnName: "value for key 3"}}
```

Reply:
> in your example, After `Next()` from key1, `key()` should return key2. `value()` should be the value of the key2 in CF2. `Next()` then will get to key3 in CF1.

I know how it works according to the current code. I was asking about @ltamasi 's proposed alternatives.

Reply:
> > Sorry, I'm not following this consistency argument either.
> 
> CF1: key1, key3 CF2: key2
> 
> MultiCfIterator with {CF1,CF2} is at key1, calls Next(), then key() returns what? If it's key2, what is value()? If it's key3, how do I get the attribute_groups()?

key2. value() and columns() would return empty since key2 is not present in CF1. attribute_groups() would return {CF2, key2's columns in CF2}.

(For key1 and key3, value() and columns() would return data from the first CF, i.e. CF1.)